[bcq-tools] Introduce preserving BCQ information tool (#3494)
Samsung/ONE: compiler/bcq-tools/README.md

# BCQ Tools

This directory includes some tools related with BCQ.

## preserve_bcq_info

### Purpose

> `preserve_bcq_info` is for preserving constant nodes which are including BCQ information.

Reviewer: (optional) `which are including ` -> `which includes`

> `preserve_bcq_info` is for preserving constant nodes which are including BCQ information.

Author: ```suggestion
`preserve_bcq_info` is for preserving constant nodes which include BCQ information.
```

When `.pb` file is converted to `.tflite` file by TFlite converter, constant nodes whose values are exactly same are removed and then linked to only one representative node.

> It causes information missing problem because we don't know which constant nodes should be linked to even we still want to apply BCQ.

Author: ```suggestion
This makes us impossible to know what constant node should be linked to a node which we want to apply BCQ.
```

> It causes information missing problem because we don't know which constant nodes should be linked to even we still want to apply BCQ.

Reviewer: maybe...
`This causes lost of information problem as we don't know which constant node should be linked to the node that we want to apply BCQ`

> Therefore, we should preserve these BCQ information.

Author: ```suggestion
One of the solutions is making all the same constant nodes different by inserting unique values and ignore the newly generated unique values when BCQ fusing is applied.
`preserve_bcq_info` will generate and insert unique dummy values to the constant nodes whose values are same not to be removed by Tensorflow Lite converter.
As a result, BCQ information will be preserved.
```
